Improve the performance of search queries on the competitions page.

pingskills opened the pull request.

larspetrus commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -87,19 +87,41 @@ def index
+        e_alias = "Events" + event_id
+        @competitions = @competitions.joins(
+          "join competition_events " + ce_alias + " ON " + ce_alias + ".competition_id = Competitions.id " \
+          "join Events " + e_alias + " ON " + e_alias + ".id = " + ce_alias + ".event_id",
```
This is a good place for interpolations:

`"join competition_events #{ce_alias} ON #{ce_alias}.competition_id = etc...`

larspetrus commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -87,19 +87,41 @@ def index
+          "join competition_events " + ce_alias + " ON " + ce_alias + ".competition_id = Competitions.id " \
+          "join Events " + e_alias + " ON " + e_alias + ".id = " + ce_alias + ".event_id",
+        ).where(e_alias + ".id = ?", event_id)
+      end
```
So if I understand the code, the main optimization here is to make more precise queries that return the exact rows we want rather than making a rough query and whittle down the selection in memory.

What evidence do we have that this actually speeds up real life queries?

jfly replied:
I'd be interested in seeing a performance analysis between this way and the old way. I suspect we'll save some time just because mysql doesn't have to allocation ~3000 ruby objects for each query.

If we do decide to go this path (pushing the work to mysql), I'd like to see all this SQL logic brought into the Competition class, rather than living out in here in the model. It's a shame that we have to implement this logic twice (once for sql, and once for ruby), but at the very least they should live in the same place.

larspetrus replied:
My suspicion is the opposite.  Some of this code, especially the multiple joins, look like it asks a lot of MySQL. Meanwhile, `.select()`-ing an array in memory should be a sub ms operation. If it's normally 3000 competitions, I agree it's crazier than I thought.

But my guess comes from a deep ignorance of the schema and data. I'm just eye balling the code. Which is why measuring - like always when optimizing - is important.

Sadly, comparing DB query speed is devilishly hard. Since the DB caches the response data, the second query you try will **always** be the fastest. By a lot. There are a number of ways to work around that.

- Turn off or invalidate the DB cache somehow. I've tried following the confident web descriptions of how to do this several times. No luck.
- Have a set of thousands search queries, throw a random 100 of them at each implementation, and compare the averages. This worked well on Birdflu.
- Deploy both as an A/B test. Collect average times over a week. Feels "dirty", but should be fairly simple and safe.

larspetrus replied:
Late night idea:

- Use two identical but separate copies of the DB.

I've dealt with this problem for many years, but that never occurred to me until I wrote this!

jfly replied:
Could you elaborate on this idea, @larspetrus? How does using 2 databases avoid the problems you listed above?

larspetrus replied:
They would have separate caches, so a query in one wouldn't affect performance in the other.

pingskills replied:
I like the ideas here and they would certainly be beneficial not only for this PR but any future performance tuning. What is the best way to implement these?

I believe it's worth putting these change in before we implement any of the above though. I don't think there will be any improvements on the initial /competitions page but anytime an event is selected, or a search is entered, we'll now be executing 30 SQL statements as opposed to over 500. Below is my analysis.

**Main page: `/competitions`**
The current implementation runs a full table scan on Competitions and then uses
`  def country
    Country.c_find(countryId)
  end`
to get each Country.

The new implementation does a full table scan of Competitions and then eager loads the Countries.
Logs show 30 SQL queries for loading the page for both versions.

**With a search: `/competitions?utf8=✓&region=all&search=Results+Barry&year=all+years&state=past&display=list`**
The current implementation does a full table scan on Competitions and then runs an additional query against `users` and `competition_delegates` for each Competition returned. It then filters the results by the search terms in Ruby.

The new implementation does a full table scan on Competitions joined to `competition_delegates`, and `users`, filters the results in the database memory to the search term.
Logs show 560 SQL statements for the original and 31 for the new

**All events selected `/competitions?utf8=✓&event_ids%5B%5D=333&event_ids%5B%5D=444&event_ids%5B%5D=555&event_ids%5B%5D=222&event_ids%5B%5D=333bf&event_ids%5B%5D=333oh&event_ids%5B%5D=333fm&event_ids%5B%5D=333ft&event_ids%5B%5D=minx&event_ids%5B%5D=pyram&event_ids%5B%5D=sq1&event_ids%5B%5D=clock&event_ids%5B%5D=skewb&event_ids%5B%5D=666&event_ids%5B%5D=777&event_ids%5B%5D=444bf&event_ids%5B%5D=555bf&event_ids%5B%5D=333mbf&region=all&search=&year=all+years&state=past&display=list`**

The current implementation does a full scan on `Competitions` and then does an extra query to `Events` joined to `competition_events` for each row.

The new implementation does a full scan on `Competitions` which is joined to `competition_events` and `Events` once for each event selected (all in this case).

Logs show 558 SQL statements for the current version and 30 for the new version.

jfly replied:
Fewer SQL statements certainly sounds better to me. That could have been ahieved with the old implementation if we had just added some `includes` though, right?

pingskills replied:
I don't think so. The current implementation retrieves selected rows from the Competitions table, and then iterates over each record passing it to a method (eg contains? for the search). I'm not great with Rails though. @larspetrus and @jonatanklosko , is there a way the current implementation could be changed to avoid the extra SQL calls without rewriting it as I've done?

jonatanklosko commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -87,19 +87,41 @@ def index
+          "%#{part}%",
+          "%#{part}%",
+          "%#{part}%",
+          "%#{part}%",
```
@larspetrus are you sure? So far I was convinced that we need these `%`

@pingskills you can use a 'variable', see [here](https://github.com/thewca/worldcubeassociation.org/blob/master/WcaOnRails/app/controllers/persons_controller.rb#L15). You can get rid of some redundancy [this way](https://github.com/thewca/worldcubeassociation.org/blob/master/WcaOnRails/app/controllers/users_controller.rb#L21-L26) if it's not too tricky in your opinion =)

jfly replied:
I agree with @jonatanklosko on both parts. It is hard to see those `%`, but they're a key part of providing this nice, expensive search feature =)

jfly commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -87,19 +87,41 @@ def index
 
     unless params[:region] == "all"
-      @competitions = @competitions.select { |competition| competition.belongs_to_region?(params[:region]) }
+      @competitions = @competitions.where("countryId = ? or continentId = ?", params[:region], params[:region])
```
Just curious... do we have indices on these two fields (`countryId` and `continentId`)? Should we?

pingskills replied:
`continentId` is on the Countries table and has an index. `countryId` is on Competitions and doesn't so we probably should add one for this.

pingskills commented:
@jfly how would you move this into the model? Would you create a `show_selected` method and basically copy across the `index` method?

jfly commented:
@pingskills, nah, I just want the SQL logic inside of the model. For example, previously we were doing something like `Competition.all.select { |c| c.belongs_to_region?(...) }`, I'd like to see the new SQL implementation as some sort of static method on the `Competition` class, something like `Competition.where_belongs_to_region(...)`.

pingskills commented:
I've now moved the SQL logic from the controller to scopes within the competition model. The contains? and belongs_to_region? methods were only used by this page so I've removed them.

jfly commented:
Awesome. Codewise, this looks good to me. However, I think if we're going to deploy it, we should have some metric of if things are actually better or worse. @pingskills, what do you think of @larspetrus's suggestions [here](https://github.com/thewca/worldcubeassociation.org/pull/1051#discussion_r93541472)?

pedrosino commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -30,6 +32,29 @@ class Competition < ActiveRecord::Base
+      Competitions.cellName like :search_term or
+      Competitions.countryId like :search_term or
+      MONTHNAME(STR_TO_DATE(Competitions.month, '%m')) like :search_term",
+      search_term: "%#{search_term}%"
```
I was discussing with Jeremy and I'm not sure if people use the month or venue search. I don't think they make much sense. Same for country, since we have an explicit country/region dropdown. I think we could remove this from here. Name and city should be enough. 
About delegate, I use the search currently, but could live with going through /admin/delegates to see someone's history. We could then use a different param (delegate_id) instead of passing it through search param. See #1035

pingskills commented:
Thanks @pedrosino. I've updated the PR to only search against the name and city.

jfly commented:
> Thanks @pedrosino. I've updated the PR to only search against the name and city.

I don't think it's a good idea to change the behavior of the code if we're looking to do a performance comparison.

pingskills commented:
Do we still think that the new relic alerts are being caused by the competitions page? If so I'd suggest we implement this PR. I can be certain that when using the search box and/or choosing events, there will be a lot less SQL statements run. If someone also selects "all years" then the current implementation will execute even more SQL statements.

I agree that having a way to test performance improvements is ideal, and perhaps we should create an issue so that someone can create such a framework for us. The problem for this PR is that I think it will be many hours of work (and I'm not sure how to do this without more research), and if the changes do cause an adverse impact it will be easy to rollback. So I'd vote for implementing as is, and monitoring the situation.

Having said that if other people think this is to risky then I'm happy to hold off until we find a good method of testing any suggested performance improvement changes. Let me know you thoughts :-)

larspetrus commented:
Well, I _want_ to approve it, I would just like to see _some_ timing data first. It doesn't have to be as complicated as I made it sound.

The vastly fewer queries must come from the `.includes(...)`. Those are no-brainers, a big improvement, and I need no convincing about them.

Making the one `@competitions` query more complex makes me a bit nervous though. Especially the added joins. In my experience, whenever you have 3 joins, you're playing with performance fire, and higher numbers are terrifying! It's very possible these joins are not like that, but I'd want to see some evidence. Or maybe they could be part of the `where` clause instead of joins.

A simple manual timing scheme is to put this line first in the controller method `@t1 = Time.now`, and this last in the view `<%= "Timing: #{Time.now - @t1}s" %>`

Just doing some averages of 5 similar queries manually should be enough to get a feel for the performance implications.


The alternative is that we make a note of the average times on New Relic the week before we deploy, and compare it a week later.

**I completely agree it's better to resolve this one way or another than to sit around waiting for perfection!**

Some kind of speed testing "framework" for this problem query should be pretty simple. We have¹ a file with 1000 real world queries. Some code that runs 20 or 50 random ones and records an average should be easy to write, and should be immune to DB caching issues.

¹ I can dig it out

pingskills commented:
Thanks @larspetrus. Below are the quick timing tests. I've shown the timing for the original code, the code with the initial changes, and the code limiting the search to only name and city. The /competitions page shows no improvement, but the others are much faster.

/competitions
**Orig:** Timing: 0.847296925s
**New:** Timing: 0.943069414s
**New with only name and city search:** Timing: 0.827037003s

/competitions?utf8=✓&region=all&search=PingSkills+Classic&state=present&year=all+years&display=list
**Orig:** Timing: 1.167606117s
**New:** Timing: 0.10751125s
**New with only name and city search:** Timing: 0.041945404s

/competitions?utf8=✓&event_ids%5B%5D=333&event_ids%5B%5D=444&event_ids%5B%5D=222&region=all&search=PingSkills+Classic&state=present&year=all+years&display=list
**Orig:** Timing: 1.144026075s
**New:** Timing: 0.19979482s
**New with only name and city search:** Timing: 0.047083485s

/competitions?utf8=✓&event_ids%5B%5D=333&event_ids%5B%5D=444&event_ids%5B%5D=222&region=all&search=My+32&year=all+years&state=past&display=list
**Orig:** Timing: 1.816054822s
**New:** Timing: 0.10397425s
**New with only name and city search:** Timing: 0.063355166s

larspetrus commented:
OK, I'm still wondering about a few things, but let's deploy this and see what happens rather than going on with this forever!

LGTM¹

Below is the performance data for the last 24 hours. Let's compare after the new code has been up for a while.

![perf](https://cloud.githubusercontent.com/assets/755243/21558261/e0a7e93c-cdeb-11e6-9a1c-b8e52776d562.png)

¹ Does someone have to say LGTM to allow committing in this repo?

jfly commented:
> ¹ Does someone have to say LGTM to allow committing in this repo?

Nope, that's just a convention I'd like us to follow for anything that isn't putting out a fire (my thoughts on this are detailed [here](https://github.com/thewca/worldcubeassociation.org/wiki/Merging-and-deploying#policy-for-merging-code)). Very open to discussing the policy if you have ideas.

larspetrus commented:
Aha. So there's no technical requirement, but an organizational one. Makes sense.

I may be confusing Github with the similar system at Google :)

I have no strong opinions on this one topic, just curious.

pingskills commented:
Deployed. Let's see if this improves things.

jfly commented:
It hasn't been a full 24 hours yet, but things look better right now:

![image](https://cloud.githubusercontent.com/assets/277474/21570251/31dccfec-ce79-11e6-9cad-8511c71dfecf.png)

(not sure if it will work in the future, but here's the newrelic link I used to grab this https://rpm.newrelic.com/accounts/1068954/applications/10385163/transactions#id=5b22436f6e74726f6c6c65722f636f6d7065746974696f6e732f696e646578222c22225d)

jfly commented:
This definitely looks like an improvement! This particular route has moved below `omni_search` in terms of most expensive queries:

![image](https://cloud.githubusercontent.com/assets/277474/21578706/84f47650-cf3f-11e6-84b9-b98d3c838acf.png)

We should tackle `PersonsController` and `omni_search` next =)

larspetrus commented:
The full 24 hour numbers look good! 

Average time is down by 60%, 499ms -> 200ms.
CompetitionsController#index is down 93%, 190ms -> 16ms

Only the _index_table.html.erb Partial remains about the same (97ms -> 88ms), which makes sense.

![grab2](https://cloud.githubusercontent.com/assets/755243/21578712/c70bef6e-cf3f-11e6-99ee-ede3321ce960.png)

pingskills commented:
Excellent. If I get a chance next week I'll take a look at
PersonsController.

Cheers,
Jeff.

On 1 January 2017 at 04:59, Lars Petrus <notifications@github.com> wrote:

> The full 24 hour numbers look good!
>
> Average time is down by 60%, 499ms -> 200ms.
> CompetitionsController#index is down 93%, 190ms -> 16ms
>
> Only the _index_table.html.erb Partial remains about the same (97ms ->
> 88ms), which makes sense.
>
> [image: grab2]
> <https://cloud.githubusercontent.com/assets/755243/21578712/c70bef6e-cf3f-11e6-99ee-ede3321ce960.png>
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/thewca/worldcubeassociation.org/pull/1051#issuecomment-269875779>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMv4HkcWUP9FyDGVw03ret3kNX6gs_Pks5rNpf6gaJpZM4LS7TU>
> .
>
